File: native/src/pathfinding.cpp
#include "vec2.hpp"
#include <vector>
#include <queue>
#include <unordered_map>
#include <cmath>

namespace woc {

/// Grid node coordinate.
struct Node {
    int c, r;
    bool operator==(const Node& o) const { return c == o.c && r == o.r; }
};

struct NodeHash {
    size_t operator()(const Node& n) const {
        return std::hash<int>()(n.c) ^ (std::hash<int>()(n.r) << 1);
    }
};

/// A* pathfinding over a flat solid grid (1 = blocked, 0 = walkable).
std::vector<Node> findPath(const std::vector<uint8_t>& solid, int width, int height,
                           Node start, Node goal) {
    auto idx = [width](int c, int r) { return r * width + c; };
    auto heuristic = [&](Node n) {
        return std::abs(n.c - goal.c) + std::abs(n.r - goal.r);
    };

    using PQItem = std::pair<int, Node>;
    std::priority_queue<PQItem, std::vector<PQItem>, std::greater<>> open;
    std::unordered_map<Node, Node, NodeHash> came;
    std::unordered_map<Node, int, NodeHash> gScore;

    open.push({0, start});
    gScore[start] = 0;

    const int dc[] = {1, -1, 0, 0};
    const int dr[] = {0, 0, 1, -1};

    while (!open.empty()) {
        Node cur = open.top().second;
        open.pop();
        if (cur == goal) break;
        for (int k = 0; k < 4; ++k) {
            Node next{cur.c + dc[k], cur.r + dr[k]};
            if (next.c < 0 || next.r < 0 || next.c >= width || next.r >= height) continue;
            if (solid[idx(next.c, next.r)] == 1) continue;
            int tentative = gScore[cur] + 1;
            if (!gScore.count(next) || tentative < gScore[next]) {
                gScore[next] = tentative;
                came[next] = cur;
                open.push({tentative + heuristic(next), next});
            }
        }
    }

    std::vector<Node> path;
    if (!came.count(goal)) return path;
    for (Node n = goal; !(n == start); n = came[n]) path.push_back(n);
    path.push_back(start);
    return {path.rbegin(), path.rend()};
}

} // namespace woc

File: src/maps/pathfinding.js
/**
 * A* pathfinding over a tilemap's solid grid.
 */
export function findPath(tilemap, start, goal) {
  const key = (c, r) => `${c},${r}`;
  const open = [{ c: start.c, r: start.r, g: 0, f: 0, parent: null }];
  const came = new Map();
  const gScore = new Map([[key(start.c, start.r), 0]]);

  const h = (c, r) => Math.abs(c - goal.c) + Math.abs(r - goal.r);

  while (open.length) {
    open.sort((a, b) => a.f - b.f);
    const cur = open.shift();
    if (cur.c === goal.c && cur.r === goal.r) {
      return reconstruct(came, cur);
    }

    for (const [dc, dr] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
      const nc = cur.c + dc;
      const nr = cur.r + dr;
      if (tilemap.tileAt(nc, nr) === -1 || tilemap.solid[nr * tilemap.width + nc] === 1) continue;
      const tentative = cur.g + 1;
      const k = key(nc, nr);
      if (tentative < (gScore.get(k) ?? Infinity)) {
        gScore.set(k, tentative);
        came.set(k, cur);
        open.push({ c: nc, r: nr, g: tentative, f: tentative + h(nc, nr), parent: cur });
      }
    }
  }
  return [];
}

function reconstruct(came, node) {
  const path = [];
  let cur = node;
  while (cur) {
    path.unshift({ c: cur.c, r: cur.r });
    cur = cur.parent;
  }
  return path;
}
